src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+        /// </summary>
+        public LoginHandler()
+        {
+            // Setup the feature extensions.
+            _sessionRecoveryFeature = new SessionRecoveryFeature();
+            _fedAuthFeature = new FedAuthFeature();
+            _tceFeature = new TceFeature();
+            _globalTransactionsFeature = new GlobalTransactionsFeature();
+            _dataClassificationFeature = new DataClassificationFeature();
+            _utf8SupportFeature = new Utf8Feature();
+            _sqlDnsCachingFeature = new SqlDnsCachingFeature();
+        }
+
+        public override async ValueTask Handle(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            LoginHandlerContext loginHandlerContext = new LoginHandlerContext(context);
+            await SendLogin(loginHandlerContext, isAsync, ct).ConfigureAwait(false);
+
+            // TODO: Complete the login by reading data. 
+            // This requires parsing by reading token stream from TDS.
+        }
+
+        private async ValueTask SendLogin(LoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+
+            PasswordChangeRequest passwordChangeRequest = context.PasswordChangeRequest;
+
+            TdsEnums.FeatureExtension requestedFeatures = TdsEnums.FeatureExtension.None;
+            FeatureExtensions features = context.Features;
+            if (context.ConnectionOptions.ConnectRetryCount > 0)
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.SessionRecovery;
+                features.SessionRecoveryRequested = true;
+            }
+
+            
+            if (ShouldRequestFedAuth(context))
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.FedAuth;
+                features.FederatedAuthenticationInfoRequested = true;
+                features.FedAuthFeatureExtensionData =
+                    new FederatedAuthenticationFeatureExtensionData
+                    {
+                        libraryType = TdsEnums.FedAuthLibrary.MSAL,
+                        authentication = context.ConnectionOptions.Authentication,
+                        fedAuthRequiredPreLoginResponse = context.FedAuthNegotiatedInPrelogin
+                    };
+            }
+
+            if (context.AccessTokenInBytes != null)
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.FedAuth;
+                features.FedAuthFeatureExtensionData = new FederatedAuthenticationFeatureExtensionData
+                {
+                    libraryType = TdsEnums.FedAuthLibrary.SecurityToken,
+                    fedAuthRequiredPreLoginResponse = context.FedAuthNegotiatedInPrelogin,
+                    accessToken = context.AccessTokenInBytes
+                };
+                // No need any further info from the server for token based authentication. So set _federatedAuthenticationRequested to true
+                features.FederatedAuthenticationRequested = true;
```

Comment:
Can you not combine both logics to create 1 method to determine whether this feature is being requested?